src/inmanta/server/services/compilerservice.py

```diff
@@ -213,37 +218,71 @@ async def run(self, force_update: Optional[bool] = False) -> Tuple[bool, Optiona
                 await self._info("Creating project directory for environment %s at %s" % (environment_id, project_dir))
                 os.mkdir(project_dir)
 
-            repo_url: str = env.repo_url
-            repo_branch: str = env.repo_branch
-            if not repo_url:
-                if not os.path.exists(os.path.join(project_dir, "project.yml")):
-                    await self._warning(f"Failed to compile: no project found in {project_dir} and no repository set")
-                await self._end_stage(0)
-            else:
-                await self._end_stage(0)
-                # checkout repo
+            async def update_modules() -> data.Report:
+                LOGGER.info("Updating modules")
+                return await self._run_compile_stage("Updating modules", inmanta_path + ["modules", "update"], project_dir)
+
+            async def install_modules() -> data.Report:
+                LOGGER.info("Installing modules")
+                return await self._run_compile_stage(
+                    "Installing modules", [*inmanta_path, "-vvv", "-X", "project", "install"], project_dir
+                )
 
-                if not os.path.exists(os.path.join(project_dir, ".git")):
+            async def setup() -> AsyncIterator[Awaitable[data.Report]]:
+                """
+                Returns an iterator over all setup stages. Inspecting stage success state is the responsibility of the caller.
+                """
+                repo_url: str = env.repo_url
+                repo_branch: str = env.repo_branch
+                if os.path.exists(os.path.join(project_dir, "project.yml")):
+                    yield self._end_stage(0)
+
+                    should_update: bool = force_update or self.request.force_update
+
+                    # switch branches
+                    if repo_branch:
+                        branch = await self.get_branch()
+                        if branch is not None and repo_branch != branch:
+                            if should_update:
+                                yield self._run_compile_stage("Fetching new branch heads", ["git", "fetch"], project_dir)
+                            yield self._run_compile_stage(
+                                f"Switching branch from {branch} to {repo_branch}",
+                                ["git", "checkout", repo_branch],
+                                project_dir,
+                            )
+                            if not should_update:
+                                # if we update, update procedure will install modules
+                                yield install_modules()
+
+                    # update project
+                    if should_update:
+                        if repo_url:
+                            yield self._run_compile_stage("Pulling updates", ["git", "pull"], project_dir)
```

Comment:
@wouterdb I'm rerequesting your review for this change in logic. It's hidden in the structural changes I made but I've added the `if repo_url` condition to perform a repo update.

Reply:
Why?

If a remote is set, isn't that sufficient?

Reply:
It is, and I guess that's a safer condition. I'll add a method to read remotes from `.git/config`.

Reply:
I decided to go one step further still and only pull if the current branch has an upstream branch associated with it.